File: Astromaximum/Classes/AmaxSettingsController.m
//
//  AmaxSettingsController.m
//  Astromaximum
//
//  Created by admin on 03.12.12.
//  Copyright (c) 2012 S&W Axis. All rights reserved.
//

#import "AmaxSettingsController.h"
#import "AmaxLocationListController.h"
#import "AmaxDataProvider.h"

@implementation AmaxSettingsController
@synthesize pickerView;

@synthesize locationListController = _locationListController;
@synthesize tvCell = _tvCell;

const int CELL_LOCATION_NAME = 0;
const int CELL_IS_CUSTOM_TIME = 1;
const int CELL_HIGHIGHT_TIME = 2;

NSString *AmaxSettingsXibNames[] = {
    @"LocationNameCell",
    @"CustomTimeSwitchCell"
};

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.title = NSLocalizedString(@"Settings", @"Settings");
        self->mDataProvider = [AmaxDataProvider sharedInstance];
    }
    return self;
}

- (void)didReceiveMemoryWarning
{
    // Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
    
    // Release any cached data, images, etc that aren't in use.
}

#pragma mark - View lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
}

- (void)viewDidUnload
{
    [self setPickerView:nil];
    [super viewDidUnload];
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [_mTableView reloadData];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    // Return YES for supported orientations
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

// Customize the number of sections in the table view.
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 2;
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *cellIdentifier = AmaxSettingsXibNames[indexPath.row];
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:cellIdentifier];
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    }
    switch ([indexPath row]) {
        case CELL_LOCATION_NAME:
            cell.textLabel.text = NSLocalizedString(@"Current location", @"Current location");
            cell.detailTextLabel.text = [mDataProvider locationName];
            break;
        case CELL_IS_CUSTOM_TIME:
            cell.textLabel.text = NSLocalizedString(@"Highlight time", @"Highlight time");
            cell.detailTextLabel.text = @"88:88";
            UISwitch *switchView = [[UISwitch alloc] initWithFrame:CGRectZero];
            cell.accessoryView = switchView;
            [switchView setOn:NO animated:NO];
            [switchView addTarget:self action:@selector(customTimeSwitchChanged:) forControlEvents:UIControlEventValueChanged];
            break;
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    if ([indexPath row] == CELL_LOCATION_NAME) {
        if (!self.locationListController) {
            self.locationListController = [[AmaxLocationListController alloc] initWithNibName:@"AmaxLocationListController" bundle:nil];
        }
        [self.navigationController pushViewController:self.locationListController animated:YES];
    }
}

- (IBAction)customTimeSwitchChanged:(id)sender
{
    UITableViewCell *targetCell = [self tableView:_mTableView cellForRowAtIndexPath:[NSIndexPath indexPathForRow:CELL_HIGHIGHT_TIME inSection:0]];
    NSLog(@"Switch changed to %@\n", [sender isOn] ? @"YES" : @"NO");
//	self.pickerView.date = [self.dateFormatter dateFromString:targetCell.detailTextLabel.text];
	
	// check if our date picker is already on screen
	if (self->pickerView.superview == nil)
	{
		[self.view.window addSubview: self->pickerView];
		
		// size up the picker view to our screen and compute the start/end frame origin for our slide up animation
		//
		// compute the start frame
		CGRect screenRect = [[UIScreen mainScreen] applicationFrame];
		CGSize pickerSize = [self->pickerView sizeThatFits:CGSizeZero];
		CGRect startRect = CGRectMake(0.0,
									  screenRect.origin.y + screenRect.size.height,
									  pickerSize.width, pickerSize.height);
		self->pickerView.frame = startRect;
		
		// compute the end frame
		CGRect pickerRect = CGRectMake(0.0,
									   screenRect.origin.y + screenRect.size.height - pickerSize.height,
									   pickerSize.width,
									   pickerSize.height);
		// start the slide up animation
		[UIView beginAnimations:nil context:NULL];
        [UIView setAnimationDuration:0.3];
		
        // we need to perform some post operations after the animation is complete
        [UIView setAnimationDelegate:self];
		
        self->pickerView.frame = pickerRect;
		
        // shrink the table vertical size to make room for the date picker
        CGRect newFrame = _mTableView.frame;
        newFrame.size.height -= self->pickerView.frame.size.height;
        _mTableView.frame = newFrame;
		[UIView commitAnimations];
		
		// add the "Done" button to the nav bar
		self.navigationItem.rightBarButtonItem = self.doneButton;
	}
}
@end

File: Astromaximum/Classes/AmaxSettingsController.swift
//
//  AmaxSettingsController.m
//  Astromaximum
//
//  Created by admin on 03.12.12.
//  Copyright (c) 2012 S&W Axis. All rights reserved.
//

import UIKit

class AmaxSettingsController : UIViewController, UITableViewDelegate, UITableViewDataSource {

    private var mDataProvider:AmaxDataProvider!
    private var mCustomTimeChanged = false
    @IBOutlet weak var _mTableView: UITableView!
    @IBOutlet weak var pickerView: UIPickerView!
    @IBOutlet weak var doneButton: UIBarButtonItem!
    @IBOutlet weak var tvCell: UITableViewCell!
    var locationListController: AmaxLocationListController!
    var helpTopicsController: AmaxHelpTopicsController!

    let CELL_LOCATION_NAME = 0
    let CELL_SHOW_CRITICAL_DEGREEES = 1
    let CELL_IS_CUSTOM_TIME = 2
    let CELL_HELP_TOPICS = 3

    let AmaxSettingsXibNames = [
        "LocationNameCell",
        "CriticalDegreesCell",
        "CustomTimeSwitchCell",
        "HelpTopicsCell",
    ]

    override init(nibName: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibName, bundle: nibBundleOrNil)
        title = NSLocalizedString("Settings", comment: "Settings")
        mDataProvider = AmaxDataProvider.sharedInstance
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func didReceiveMemoryWarning() {
        // Releases the view if it doesn't have a superview.
        super.didReceiveMemoryWarning()

        // Release any cached data, images, etc that aren't in use.
    }

    // MARK: - View lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view from its nib.
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        mCustomTimeChanged = false
        _mTableView.reloadData()
    }

    override func viewDidDisappear(_ animated:Bool) {
        if mCustomTimeChanged {
            if let topVC = UIApplication.shared.topMostViewController() as? AmaxPageController {
                topVC.currentController().cachedStartTime = -1  // to force update
                _ = topVC.currentController().updateDisplay()
            }
        }
    }

    // Customize the number of sections in the table view.
    func numberOfSectionsInTableView(tableView: UITableView!) -> Int {
        return 1
    }

    func tableView(_ tableView:UITableView, numberOfRowsInSection section:Int) -> Int {
        return 4
    }

    // Customize the appearance of table view cells.
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = AmaxSettingsXibNames[indexPath.row]
        var cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .value1, reuseIdentifier: cellIdentifier)
            cell?.accessoryType = .disclosureIndicator
        }
        guard let c = cell else {
            return cell!
        }
        switch (indexPath.row) {
            case CELL_LOCATION_NAME:
                c.textLabel?.text = NSLocalizedString("Current_location", comment: "Current location")
                c.detailTextLabel?.text = mDataProvider.locationName()
            case CELL_SHOW_CRITICAL_DEGREEES:
                c.textLabel?.text = NSLocalizedString("pref_critical_degrees", comment: "")
                let switchView = UISwitch(frame: CGRect.zero)
                c.accessoryView = switchView
                switchView.setOn(mDataProvider.mShowCriticalDegrees, animated: false)
                switchView.addTarget(self, action: #selector(self.criticalDegreesSwitchChanged(sender:)), for: .valueChanged)
            case CELL_HELP_TOPICS:
                c.textLabel?.text = NSLocalizedString("help_title", comment: "Help list")
            case CELL_IS_CUSTOM_TIME:
                c.textLabel?.text = NSLocalizedString("Highlight_time", comment: "Highlight time")
                c.detailTextLabel?.text = String(format: "%02d:%02d",
                                                 AmaxDataProvider.sharedInstance.mCustomHour,
                                                 AmaxDataProvider.sharedInstance.mCustomMinute)

                let switchView = UISwitch(frame: CGRect.zero)
                c.accessoryView = switchView
                switchView.setOn(mDataProvider.mUseCustomTime, animated: false)
                switchView.addTarget(self, action: #selector(self.customTimeSwitchChanged(sender:)), for: .valueChanged)
            default:
                break
        }
        return cell!
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        switch(indexPath.row) {
        case CELL_LOCATION_NAME:
            if locationListController == nil {
                locationListController = AmaxLocationListController(nibName: "AmaxLocationListController", bundle: nil)
            }
            self.navigationController?.pushViewController(locationListController, animated: true)
        case CELL_HELP_TOPICS:
            if helpTopicsController == nil {
                helpTopicsController = AmaxHelpTopicsController(nibName: "AmaxHelpTopicsController", bundle: nil)
            }
            self.navigationController?.pushViewController(helpTopicsController, animated: true)
        case CELL_IS_CUSTOM_TIME:
            let targetCell = self.tableView(_mTableView, cellForRowAt: IndexPath(row: indexPath.row, section: 0))
            let sw = targetCell.accessoryView as! UISwitch
            if sw.isOn {
                showCustomTimeDialog(cell: targetCell, keepOn: true)
            }
        default:
            break
        }
    }

    @IBAction func criticalDegreesSwitchChanged(sender: AnyObject!) {
        mDataProvider.mShowCriticalDegrees = sender.isOn
        mDataProvider.saveCurrentState()
    }

    @IBAction func customTimeSwitchChanged(sender: AnyObject!) {
        let targetCell = tableView(_mTableView, cellForRowAt: IndexPath(row: CELL_IS_CUSTOM_TIME, section: 0))
        if sender.isOn {
            showCustomTimeDialog(cell: targetCell, keepOn: false)
        }
        else {
            mDataProvider.mUseCustomTime = false
            mCustomTimeChanged = true
            mDataProvider.saveCurrentState()
        }
    }
    
    func showCustomTimeDialog(cell: UITableViewCell, keepOn: Bool) {
        let dpd = DatePickerDialog()
        dpd.show(  NSLocalizedString("time_picker_title", comment:"")
                 , doneButtonTitle: NSLocalizedString("date_picker_done", comment:"")
                 , cancelButtonTitle: NSLocalizedString("date_picker_cancel", comment:"")
                 , defaultDate: mDataProvider.getCustomTimeForPicker()
                 , datePickerMode: .time) { time in
            if let tm = time {
                self.mDataProvider.setCustomTime(from: tm)
                self.mDataProvider.mUseCustomTime = true
            }
            else {
                if !keepOn {
                    self.mDataProvider.mUseCustomTime = false
                }
            }
            self._mTableView.reloadData()
            self.mCustomTimeChanged = true
            self.mDataProvider.saveCurrentState()
        }
    }
}
